### PR TITLE
Menubar tests

### PR DESCRIPTION
**Pull request by Crazylemon64**
This is currently a complete skeleton of a test suite for the menubar. I'm having trouble actually inspecting the contents of the menubar's state, but then again maybe I'm looking at this absolutely boneheadedly.

### CODE REVIEW

**Comment by Elv13** [comment]
Thanks a lot for this. This is the error:

```
+DISPLAY=:5 /home/travis/build/awesomeWM/awesome/tests/../utils/awesome-client

+cat /home/travis/build/awesomeWM/awesome/tests/test-menubar.lua

+tail -n 100000 -f --pid 22942 /tmp/tmp.ePnqulQA2B/_awesome_test.log

+sleep 120

2016-09-07 23:08:20 W: awesome: a_glib_poll:373: Last main loop iteration took 1.083486 seconds! Increasing limit for this warning to that value.

Test finished successfully

2016-09-07 23:08:25 W: awesome: a_glib_poll:373: Last main loop iteration took 4.262665 seconds! Increasing limit for this warning to that value.

2016-09-07 23:08:25 W: awesome: luaA_panic:420: unprotected error in call to Lua API ((null))

2016-09-07 23:08:25 W: awesome: luaA_panic:423: dumping backtrace

/home/travis/build/awesomeWM/awesome/tests/../build/awesome(backtrace_get+0x41) [0x41ff9d]

/home/travis/build/awesomeWM/awesome/tests/../build/awesome() [0x416426]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xadea) [0x7fd1c35d9dea]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x17677) [0x7fd1c35e6677]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xb989) [0x7fd1c35da989]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(lua_callk+0x38) [0x7fd1c35d6bc8]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xb254) [0x7fd1c35da254]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x17631) [0x7fd1c35e6631]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xb989) [0x7fd1c35da989]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x156ea) [0x7fd1c35e46ea]

/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x15a18) [0x7fd1c35e4a18]

/usr/local/lib/lua/5.2/lgi/corelgilua51.so(lgi_type_get_repotype+0xdf) [0x7fd1be75185f]

/usr/local/lib/lua/5.2/lgi/corelgilua51.so(+0x11a32) [0x7fd1be757a32]

/usr/local/lib/lua/5.2/lgi/corelgilua51.so(lgi_object_2lua+0xc1) [0x7fd1be7585a1]

/usr/local/lib/lua/5.2/lgi/corelgilua51.so(lgi_marshal_2lua+0x30f) [0x7fd1be75649f]

/usr/local/lib/lua/5.2/lgi/corelgilua51.so(+0x8a10) [0x7fd1be74ea10]

/usr/local/lib/lua/5.2/lgi/corelgilua51.so(+0x9179) [0x7fd1be74f179]

/usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_closure_unix64_inner+0x1fb) [0x7fd1c22238cb]

/usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_closure_unix64+0x46) [0x7fd1c2223c44]

/usr/lib/x86_64-linux-gnu/libgio-2.0.so.0(+0x808db) [0x7fd1c26c48db]

/usr/lib/x86_64-linux-gnu/libgio-2.0.so.0(+0x808f9) [0x7fd1c26c48f9]

/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_dispatch+0x135) [0x7fd1c5e25ce5]

/lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x49048) [0x7fd1c5e26048]

/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_loop_run+0x6a) [0x7fd1c5e2630a]

/home/travis/build/awesomeWM/awesome/tests/../build/awesome(main+0x1054) [0x40f97f]

/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7fd1c2fe6f45]

/home/travis/build/awesomeWM/awesome/tests/../build/awesome() [0x40e389]

2016-09-07 23:08:25 W: awesome: luaA_panic:424: restarting awesome

2016-09-07 23:08:27 W: awesome: a_glib_poll:373: Last main loop iteration took 1.209394 seconds! Increasing limit for this warning to that value.

+ps -o comm= 21426

+[  = awesome ]

+ps -o comm= 21537

+[  = awesome ]

+ps -o comm= 21669

+[  = awesome ]

+ps -o comm= 21771

+[  = awesome ]

+ps -o comm= 21897

+[  = awesome ]

+ps -o comm= 22008

+[  = awesome ]

+ps -o comm= 22241

+[  = awesome ]

+ps -o comm= 22472

+[  = awesome ]

+ps -o comm= 22643

+[  = awesome ]

+ps -o comm= 22754

+[  = awesome ]

+ps -o comm= 22942

+[ awesome = awesome ]

+echo Killing (stale?!) awesome (PID 22942) after 120 seconds.

Killing (stale?!) awesome (PID 22942) after 120 seconds.

+kill 22942

(process:22942): GLib-WARNING **: poll(2) failed due to: Resource temporarily unavailable.

+grep -q -E ^Test finished successfully$ /tmp/tmp.ePnqulQA2B/_awesome_test.log

+grep -q -E [Ee]rror|assertion failed /tmp/tmp.ePnqulQA2B/_awesome_test.log

+echo ===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-menubar.lua! <===

===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-menubar.lua! <===

+grep --color -o --binary-files=text -E .*[Ee]rror.*|.*assertion failed.* /tmp/tmp.ePnqulQA2B/_awesome_test.log

2016-09-07 23:08:25 W: awesome: luaA_panic:420: unprotected error in call to Lua API ((null))
```

I wonder if this is actually true:

    (process:22942): GLib-WARNING **: poll(2) failed due to: Resource temporarily unavailable.

In that case, all the Awesome instances (what the hell, why is there so many of them!) had eaten up all the memory. Umm, Awesome later crashed because it didn't handle that (I just tested spawing more process than the limit and opening more file descriptor than the system max per process and it didn't crash this way. Actually getting a real out of memory error would crash my PC (thanks Intel GPU driver!...), so I didn't tried.

Anyway, this error is probably the consequence. Also, one of the target did build. And finally the multi-screen stuff I just merged probably made things less stable, but the tests did pass when I tried it.

File: tests/test-desktop-files.lua [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+local assets_dir = os.getenv("TEST_ASSETS") .. "/"
+menubar.menu_gen.all_menu_dirs = { assets_dir }
+
+runner.run_steps({
+    function()
+        print("Running desktop file test!")
+        menubar.menu_gen.generate(function(menu_entries)
+            print("Menu list generated!")
+            gears_debug.dump(menu_entries, "menu_entries", 4)
+        end)
```
Comment (Crazylemon64) [inline tests/test-desktop-files.lua]: This bit's async - meaning the code here isn't getting run by the time the test finishes. Is there some way around this?

Comment (Elv13) [inline tests/test-desktop-files.lua]: The test are executed up to 5 time if they return nil. In case you have something async, just return nil and it will be executed again 100ms later. So, execute the code, return true without further check and in the next test step, check the code. Do not assert, just return nil. If the test don't return true after the last attempt, it will be considered a failure anyway.

**Comment by blueyed** [comment]
@Crazylemon64 
What's the state of this?  Would you like to push it further and finish it (with a little help from your friends)?

**Comment by Crazylemon64** [comment]
Ah, yes! I've been caught up with other things in my life, but at this point I think I finally have a good expanse of time where I can sit down and polish this off.

**Comment by Crazylemon64** [comment]
Alright! There should be a non-zero quantity of tests present now!

How does it look?

File: tests/test-desktop-files.lua [hunk 2]
```diff
@@ -0,0 +1,34 @@
+local runner = require("_runner")
```
Comment (Elv13) [inline tests/test-desktop-files.lua]: I don't think this test has to be an integration one. Given you control the files, it could be a `spec/` one. They are much faster.

**Comment by Elv13** [comment]
It crashes Awesome:

```
== Running /home/travis/build/awesomeWM/awesome/tests/test-desktop-files.lua ==
2017-04-01 06:01:30 W: awesome: a_glib_poll:389: Last main loop iteration took 0.867668 seconds! Increasing limit for this warning to that value.
2017-04-01 06:02:12 W: awesome: a_glib_poll:389: Last main loop iteration took 41.271526 seconds! Increasing limit for this warning to that value.
2017-04-01 06:02:12 W: awesome: luaA_panic:529: unprotected error in call to Lua API (what)
2017-04-01 06:02:12 W: awesome: luaA_panic:532: dumping backtrace
/home/travis/build/awesomeWM/awesome/build/awesome(backtrace_get+0x41) [0x421066]
/home/travis/build/awesomeWM/awesome/build/awesome() [0x416e36]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xadea) [0x7f26c5eefdea]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x17677) [0x7f26c5efc677]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xb989) [0x7f26c5ef0989]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(lua_callk+0x38) [0x7f26c5eecbc8]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xb254) [0x7f26c5ef0254]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x17631) [0x7f26c5efc631]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xb989) [0x7f26c5ef0989]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x156ea) [0x7f26c5efa6ea]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x15a18) [0x7f26c5efaa18]
/usr/local/lib/lua/5.2/lgi/corelgilua51.so(lgi_type_get_repotype+0xdf) [0x7f26c0e6285f]
/usr/local/lib/lua/5.2/lgi/corelgilua51.so(+0x11a32) [0x7f26c0e68a32]
/usr/local/lib/lua/5.2/lgi/corelgilua51.so(lgi_object_2lua+0xc1) [0x7f26c0e695a1]
/usr/local/lib/lua/5.2/lgi/corelgilua51.so(lgi_marshal_2lua+0x30f) [0x7f26c0e6749f]
/usr/local/lib/lua/5.2/lgi/corelgilua51.so(+0x8a10) [0x7f26c0e5fa10]
/usr/local/lib/lua/5.2/lgi/corelgilua51.so(+0x9179) [0x7f26c0e60179]
/usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_closure_unix64_inner+0x1fb) [0x7f26c4b368cb]
/usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_closure_unix64+0x46) [0x7f26c4b36c44]
/usr/lib/x86_64-linux-gnu/libgio-2.0.so.0(+0x808db) [0x7f26c4fd78db]
/usr/lib/x86_64-linux-gnu/libgio-2.0.so.0(+0x808f9) [0x7f26c4fd78f9]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_dispatch+0x135) [0x7f26c873fce5]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x49048) [0x7f26c8740048]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_loop_run+0x6a) [0x7f26c874030a]
/home/travis/build/awesomeWM/awesome/build/awesome(main+0x111f) [0x40ff32]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7f26c58f9f45]
/home/travis/build/awesomeWM/awesome/build/awesome() [0x40e689]
2017-04-01 06:02:12 W: awesome: luaA_panic:533: restarting awesome
2017-04-01 06:02:14 W: awesome: a_glib_poll:389: Last main loop iteration took 0.928271 seconds! Increasing limit for this warning to that value.
```

> 2017-04-01 06:02:12 W: awesome: a_glib_poll:389: Last main loop iteration took 41.271526 
> seconds! Increasing limit for this warning to that value.

At first sight, it looks like an infinite something until it explodes.

**Comment by Crazylemon64** [comment]
oh. oh dear. Alright, I'll see what's different on my machine that made this work there and not here

**Comment by psychon** [comment]
> It crashes Awesome:

Menubar uses `require("lgi").Gio.Async.start()`, so it can cause use-after-free-run in any LGI version without https://github.com/pavouk/lgi/commit/eb40be487c2f3f94dafac3c5070104559139b00a. However, no version of LGI with that fix was released yet.

The symptom matches the above. The Lua 5.3 build (with Lua-internal assertion enabled) fails with
```
awesome: lapi.c:915: lua_callk: Assertion `(L->status == 0) && "cannot do calls on non-normal thread"' failed.
```
So ignore the crash. The only 'real'  problem is the assertion failuire (I didn't look at what the assertion is, but we should not forget that the above issue can basically manifest itself in random ways):
```
Error: running function for step 2/2 (@2): [string "local runner = require("_runner")..."]:27: assertion failed!
```

File: tests/test-desktop-files.lua [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        if menu_gen_list == nil then
+            -- Continue waiting
+            if count >= 5 then
+                return false
+            end
+            return
+        else
+            -- Begin the actual test
+            local active_entry = menu_gen_list[1] -- Lua array counting starts at 1, right?
+            assert(active_entry.icon == "/usr/share/icons/hicolor/128x128/apps/firefox.png")
```
Comment (psychon) [inline tests/test-desktop-files.lua]: What's the first entry of this table supposed to be? I don't see such a .desktop file being added anywhere in this commit. Does the test only work if Firefox is installed? Well, Firefox is not installed on Travis.

Comment (psychon) [inline tests/test-desktop-files.lua]: Oh and, the following would produce a (only slightly) better error message:
```lua
assert(active_entry.icon == "/usr/share/icons/hicolor/128x128/apps/firefox.png", active_entry.icon)
```

Comment (Crazylemon64) [inline tests/test-desktop-files.lua]: It's peppered across several commits - I'll re-squash commits around so things make more sense, but the file itself is [here](https://github.com/Crazylemon64/awesome/blob/be7d02b56ea4d18387ca55d299d2e0a0b02ae116/tests/assets/example_1.desktop). Last I looked over icon path generation code, it didn't actually require any icons to be installed on the system, but I'll give another look-over to make certain of that.

Comment (psychon) [inline tests/test-desktop-files.lua]: That file only says `Icon=firefox`, something will have to turn that into a full path...?

**Comment by Elv13** [comment]
@Crazylemon64 ping, one final push so the tests pass and the PR can be merged? It doesn't have to cover every issues linked in the header, just some tests are already better than none.

**Comment by blueyed** [comment]
@Crazylemon64 
Ping again.
Can you please update us?

**Comment by Crazylemon64** [comment]
I set the `XDG_CACHE_DIR` variable to the `tmp_files` directory - this way instead of getting `/dev/null/.cache/awesome/menu_count_file`, we put it inside the directory that we just clean up anyways.

Though, I did remove the clause that blocked testing on Lua 5.3, as I use that on my own machine, and running the tests there worked out. Should I put that check back in...?

**Comment by psychon** [comment]
> Though, I did remove the clause that blocked testing on Lua 5.3, as I use that on my own machine, and running the tests there worked out. Should I put that check back in...?

Feel free to put it back in if it works. As far as I remember, it also works locally for me and on Travis. However, it fails under Travis when coverage was being measured. I'm all for fixing this stuff instead of hiding its brokeness.

File: spec/menubar/desktop_files_spec.lua [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            assert(active_entry.icon == nil, active_entry.icon)
+            assert(active_entry.cmdline == "ls", active_entry.cmdline)
+            -- If we want to make things show up in multi categories this should be a full table
+            gdebug.dump(active_entry.category, "Category", 4)
+            assert(active_entry.category == "games", active_entry.category)
+            assert(active_entry.name == "Example", active_entry.name)
+            return true
+        end
+    end,
+})
```
Comment (blueyed) [inline spec/menubar/desktop_files_spec.lua]: This looks like it should go into `tests/`.

File: tests/run.sh [hunk 5]
```diff
@@ -92,10 +93,12 @@ cleanup() {
         kill -TERM "$p" 2>/dev/null || true
     done
     rm -rf "$tmp_files" || true
+    rm -rf
```
Comment (blueyed) [inline tests/run.sh]: wtf?

File: spec/menubar/desktop_files_spec.lua (hunk 4 above)
Comment (blueyed) [inline spec/menubar/desktop_files_spec.lua]: ..or rather become a real busted test (which would be better, since it is faster).

File: tests/run.sh (hunk 5 above)
Comment (Crazylemon64) [inline tests/run.sh]: oh no, I added a thing which I then forgot about when I realized I could use the `tmp_files` thing - and I forgot to remove this half-implemented stub

**Comment by Crazylemon64** [comment]
https://gist.github.com/Crazylemon64/953710bca67720e76041bea7ae39f431

Alright, I've been working on converting this to a busted test, but I'm stuck at the point of making it cooperate with the async menu entry generation. It states that `async` is not defined at the moment of calling, <a href=http://olivinelabs.com/busted/#async-tests>even though busted documentation states that it is defined here</a>. Any advice?

**Comment by Crazylemon64** [comment]
ohhhhh boy this killed travis, in a very grisly way
this testing is going to need a lot of refinement
```
lua: lapi.c:915: lua_callk: Assertion `(L->status == 0) && "cannot do calls on non-normal thread"' failed.

Aborted (core dumped)
```

**Comment by Crazylemon64** [comment]
https://github.com/Crazylemon64/awesome/blob/111bc9e570833caa5e18b42759f182b1a6e88423/spec/menubar/desktop_files_spec.lua
The file that caused travis to explode, for your convenience

**Comment by psychon** [comment]
Ignore that failed assertion. The bug is https://github.com/pavouk/lgi/issues/155 and we are trying to talk @pavouk into doing a new release. That would fix this failed assertion. We cannot do anything about it on our end.

**Comment by Crazylemon64** [comment]
Looking [here](https://github.com/pavouk/lgi/releases), it seems that new version of LGI was released, right? Should I resume work on this, then?

**Comment by psychon** [comment]
I tried restarting some of the older Travis builds for this PR to see what happens with a newer LGI version. Well...

The Lua 5.3 job failed with
```
[  1%] Building C object CMakeFiles/lgi-check.dir/build-utils/lgi-check.c.o
Linking C executable lgi-check
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../../lib/liblua.so: undefined reference to `dlopen'
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../../lib/liblua.so: undefined reference to `dlclose'
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../../lib/liblua.so: undefined reference to `dlerror'
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../../lib/liblua.so: undefined reference to `dlsym'
```
Lua 5.1 failed with
```
== Running /home/travis/build/awesomeWM/awesome/tests/test-menubar.lua ==
2017-11-10 20:54:29 W: awesome: a_glib_poll:432: Last main loop iteration took 1.262278 seconds! Increasing limit for this warning to that value.
2017-11-10 20:54:29 W: awesome: luaA_panic:541: unprotected error in call to Lua API (attempt to call a number value)
2017-11-10 20:54:29 W: awesome: luaA_panic:544: dumping backtrace
[...]
```
Lua 5.2... uhm...
```
Error → spec/menubar/desktop_files_spec.lua @ 11
menubar.menu_gen Entry Generation setup
/usr/local/lib/luarocks/rocks/busted/2.0.rc12-1/bin/busted:3: true
make[4]: *** [CMakeFiles/check-unit-coverage] Error 1
```
So, feel free to continue working on this, but it doesn't look all that simple...

**Comment by Elv13** [comment]
Try to rebase the branch first

**Comment by Sorky** [comment]
Some test updates that may overlap or give ideas

https://github.com/awesomeWM/awesome/pull/2601
